tide/core/TiffPyramidReader.cpp

```diff
@@ -0,0 +1,155 @@
+/*                                                                   */
+/*    THIS  SOFTWARE IS PROVIDED  BY THE  UNIVERSITY OF  TEXAS AT    */
+/*    AUSTIN  ``AS IS''  AND ANY  EXPRESS OR  IMPLIED WARRANTIES,    */
+/*    INCLUDING, BUT  NOT LIMITED  TO, THE IMPLIED  WARRANTIES OF    */
+/*    MERCHANTABILITY  AND FITNESS FOR  A PARTICULAR  PURPOSE ARE    */
+/*    DISCLAIMED.  IN  NO EVENT SHALL THE UNIVERSITY  OF TEXAS AT    */
+/*    AUSTIN OR CONTRIBUTORS BE  LIABLE FOR ANY DIRECT, INDIRECT,    */
+/*    INCIDENTAL,  SPECIAL, EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES    */
+/*    (INCLUDING, BUT  NOT LIMITED TO,  PROCUREMENT OF SUBSTITUTE    */
+/*    GOODS  OR  SERVICES; LOSS  OF  USE,  DATA,  OR PROFITS;  OR    */
+/*    BUSINESS INTERRUPTION) HOWEVER CAUSED  AND ON ANY THEORY OF    */
+/*    LIABILITY, WHETHER  IN CONTRACT, STRICT  LIABILITY, OR TORT    */
+/*    (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING IN ANY WAY OUT    */
+/*    OF  THE  USE OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE    */
+/*    POSSIBILITY OF SUCH DAMAGE.                                    */
+/*                                                                   */
+/* The views and conclusions contained in the software and           */
+/* documentation are those of the authors and should not be          */
+/* interpreted as representing official policies, either expressed   */
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#include "TiffPyramidReader.h"
+
+#include "log.h"
+#include "types.h"
+
+#include <tiffio.h>
+
+struct TIFFDeleter
+{
+    void operator()( TIFF* file ) { TIFFClose( file ); }
+};
+typedef std::unique_ptr<TIFF, TIFFDeleter> TIFFPtr;
+
+struct TiffPyramidReader::Impl
+{
+    Impl( const QString& uri )
+        : tif{ TIFFOpen( uri.toLocal8Bit().constData(), "r" )}
+    {
+        if( !tif )
+            throw std::runtime_error( "File could not be opened" );
+
+        if( !TIFFIsTiled( tif.get( )))
+            throw std::runtime_error( "Not a tiled tiff image" );
+    }
+    TIFFPtr tif;
+};
+
+TiffPyramidReader::TiffPyramidReader( const QString& uri )
+    : _impl{ new Impl{ uri }}
+{}
+
+TiffPyramidReader::~TiffPyramidReader() {}
+
+QSize TiffPyramidReader::getImageSize() const
+{
+    QSize size;
+    TIFFGetField( _impl->tif.get(), TIFFTAG_IMAGEWIDTH, &size.rwidth( ));
+    TIFFGetField( _impl->tif.get(), TIFFTAG_IMAGELENGTH, &size.rheight( ));
+    return size;
+}
+
+QSize TiffPyramidReader::getTileSize() const
+{
+    QSize size;
+    TIFFGetField( _impl->tif.get(), TIFFTAG_TILEWIDTH, &size.rwidth( ));
+    TIFFGetField( _impl->tif.get(), TIFFTAG_TILELENGTH, &size.rheight( ));
+    return size;
+}
+
+int TiffPyramidReader::getBytesPerPixel() const
+{
+    int value = 0;
+    TIFFGetField( _impl->tif.get(), TIFFTAG_SAMPLESPERPIXEL, &value );
+    return value;
+}
+
+int TiffPyramidReader::findTopPyramidLevel()
```

Comment:
not really because the state of the TIFF* is modifed by commands such as TIFFSetDirectory() which are required to read from a different LOD. Unfortunately readTile() has the same issue, they cannot be const and are not thread-safe...